xdsl/dialects/arith.py

```diff
@@ -140,31 +140,41 @@ def from_float_and_width(
             val = FloatAttr(val, typ)
         return Constant.create(result_types=[typ], attributes={"value": val})
 
-
-class BinaryOperation(IRDLOperation):
+_T = TypeVar("_T")
+class BinaryOperation(IRDLOperation, Generic[_T]):
     """A generic operation. Operation definitions inherit this class."""
 
+    traits = frozenset([Pure()])
+
+    lhs: Annotated[Operand, _T]
+    rhs: Annotated[Operand, _T]
+    result: Annotated[OpResult, _T]
+
+    @classmethod
+    def get(
+        cls,
+        operand1: Union[Operation, SSAValue], operand2: Union[Operation, SSAValue]
```

Comment:
Instead of `Union[A, B]` i would suggest `A | B`. It's more in line with the rest of the repo.